src/include/concurrency/lock_manager.h

```diff
@@ -0,0 +1,147 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// lock_manager.h
+//
+// Identification: src/include/concurrency/lock_manager.h
+//
+// Copyright (c) 2015-18, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <map>
+#include <utility>
+
+#include "common/internal_types.h"
+#include "common/logger.h"
+#include <boost/thread/shared_mutex.hpp>
+#include <boost/foreach.hpp>
+
+namespace peloton {
+namespace concurrency {
+
+//===--------------------------------------------------------------------===//
+// Lock manager
+//===--------------------------------------------------------------------===//
+
+class LockManager {
+ public:
+
+  //===--------------------------------------------------------------------===//
+  // Safe lock
+  //===--------------------------------------------------------------------===//
+  // Initial one of these after initialed lock on oid
+  // Remember to initial it as local variable
+  // It will automatically unlock the corresponding lock when the current scope
+  // exits (since the unlock logic is in its destructor).
+  class SafeLock {
+   public:
+    // Type of locked read write locks.
+    enum RWLockType { SHARED, EXCLUSIVE, INVALID };
+
+    // Default Constructor
+    SafeLock(){
+      oid_ = INVALID_OID;
+      type_ = RWLockType::INVALID;
+    };
+
+    SafeLock(oid_t oid, RWLockType type){
+      oid_ = oid;
+      type_ = type;
+    }
+
+    // Set the oid and type for the lock
+    void Set(oid_t oid, RWLockType type){
+      oid_ = oid;
+      type_ = type;
+    }
+
+    // Unlock the given lock in the constructor
+    ~SafeLock(){
+      LockManager *lm = LockManager::GetInstance();
+      if (type_ == RWLockType::EXCLUSIVE){
+        lm->UnlockExclusive(oid_);
+      }
+      else if (type_ == RWLockType::SHARED){
+        lm->UnlockShared(oid_);
+      }
+    }
+
+   // Members
+   private:
+    oid_t oid_;
+    RWLockType type_;
+  };
+
+  // Type of locks. Currently only support RWLOCK
+  enum LockType { RW_LOCK };
+
+  // Constructor
+  LockManager() {}
```

Comment:
If we use global singleton, it's better to put constructor into private region.

Reply:
Could you specify what is a singleton in your opinion?